mh/state.py

```diff
@@ -589,9 +792,518 @@ def match_city(city, fields):
                 ])
         return cities
 
+    def update_players(self):
+        # Central server method for clearing unused Capcom IDs
+        for capcom_id, player_info in self.capcom_ids.items():
+            if player_info["session"] is not None and \
+                player_info["session"].local_info["server_id"] and \
+                capcom_id not in self.cache.players:
+                self.capcom_ids[capcom_id]["session"] = None
+            elif capcom_id in self.cache.players:
+                self.capcom_ids[capcom_id]["session"] = self.cache.players[capcom_id]
+
+    def update_capcom_id(self, session):
+        # Central server method for keeping track of in-use Capcom IDs
+        self.capcom_ids[session.capcom_id]["session"] = session
+
 
 CURRENT_STATE = State()
 
 
 def get_instance():
     return CURRENT_STATE
+
+
+class PacketTypes:
+    FriendlyHello = 0x0001
+    ReqConnectionInfo = 0x0002
+    SendConnectionInfo = 0x0003
+    ReqServerRefresh = 0x0004
+    SessionInfo = 0x0005
+
+
+class CentralConnectionHandler(object):
+    def __init__(self, socket, client_address, cache):
+        self.id = 0
+        self.socket = socket
+        self.client_address = client_address
+        self.cache = cache
+        self.rfile = self.socket.makefile('rb', -1)
+        self.wfile = self.socket.makefile('wb', 0)
+
+        self.rw = Lock()
+        self.finished = False
+
+        self.handler_functions = {
+            PacketTypes.FriendlyHello: self.RecvFriendlyHello,
+            PacketTypes.SendConnectionInfo: self.RecvConnectionInfo,
+            PacketTypes.ReqConnectionInfo: self.RecvReqConnectionInfo,
+            PacketTypes.ReqServerRefresh: self.RecvReqServerRefresh,
+            PacketTypes.SessionInfo: self.RecvSessionInfo,
+        }
+
+    def fileno(self):
+        # type: () -> int
+        return self.socket.fileno()
+
+    def on_recv(self):
+        header = self.rfile.read(10)
+        if not len(header) or len(header) < 10:
+            return None
+
+        return self.recv_packet(header)
+
+    def recv_packet(self, header):
+        size, packet_id, server_id = struct.unpack(">IIH", header)
+        data = self.rfile.read(size)
+        return server_id, packet_id, data
+
+    def send_packet(self, packet_id=0, data=b""):
+        self.wfile.write(self.pack_data(
+            data, packet_id
+        ))
+
+    def pack_data(self, _data, packet_id):
+        data = struct.pack(">II", len(_data), packet_id)
+        data += _data
+        return data
+
+    def is_finished(self):
+        return self.finished
+
+    def on_exception(self, e):
+        self.finish()
+
+    def find_handler(self, packet):
+        self.handler_functions[packet[1]](packet[0], packet[2])
```

Comment:
`packet` should be unpacked into variables with appropriate names.

Reply:
This has been fixed